source/scheduler.hpp

```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include <queue>
+
+#include "event.hpp"
+
+namespace sim {
+
+// Scheduler is implemented as a Singleton class
+// which provides a global access to a single instance
+class Scheduler {
+public:
+    // Static method to get the instance
+    static Scheduler& get_instance() {
+        static Scheduler instance;
+        return instance;
+    }
+
+    void tick();
+    void add(Event& event);
+
+private:
+    // Private constructor to prevent instantiation
+    Scheduler() {}
+    // No copy constructor and assignment operators
+    Scheduler(const Scheduler&) = delete;
+    Scheduler& operator=(const Scheduler&) = delete;
+
+    std::priority_queue<Event*> m_events;
```

Comment:
I think pointers is redundant here; moreover now we need to add comparator for `Event*` type. Easier solution:
```suggestion
    std::priority_queue<Event> m_events;
```